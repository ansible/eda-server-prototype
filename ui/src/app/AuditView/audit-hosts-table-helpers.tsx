import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import { Text, TextVariants } from '@patternfly/react-core';

export const createRows = (data) =>
  data.map(({ host, rule, ruleset, fired_date }) => ({
    cells: [
      host,
      <Fragment key={`[audit-host-${rule?.id}`}>
        <Link
          to={{
            pathname: `/rule/${rule?.id}`,
          }}
        >
          {rule?.name || rule?.id}
        </Link>
      </Fragment>,
      <Fragment key={`[audit-ruleset-${ruleset?.id}`}>
        <Link
          to={{
            pathname: `/ruleset/${ruleset?.id}`,
          }}
        >
          {ruleset?.name || ruleset?.id}
        </Link>
      </Fragment>,
      <Fragment key={`[audit-last_fired-${rule?.name}`}>
        <Text component={TextVariants.small}>
          {new Intl.DateTimeFormat('en-US', { dateStyle: 'short', timeStyle: 'long' }).format(
            new Date(fired_date || 0)
          )}
        </Text>
      </Fragment>,
    ],
  }));
