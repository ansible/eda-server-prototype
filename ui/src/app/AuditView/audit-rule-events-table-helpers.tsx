import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import {Text, TextVariants} from "@patternfly/react-core";

export const createRows = (data) =>
  data.map(({ id, name, increment_counter, source_type, timestamp }) => ({
    id,
    cells: [
      <Fragment key={`[event-${id}`}>
        <Link
          to={{
            pathname: `/event/${id}`,
          }}
        >
          {name || id}
        </Link>
      </Fragment>,
      increment_counter,
      source_type,
      <Fragment key={`[audit-rule-${id}-event`}>
        <Text component={TextVariants.small}>
          {new Intl.DateTimeFormat('en-US', { dateStyle: 'short', timeStyle: 'long' }).format(
            new Date(timestamp || 0)
          )}
        </Text>
      </Fragment>,
    ],
  }));
