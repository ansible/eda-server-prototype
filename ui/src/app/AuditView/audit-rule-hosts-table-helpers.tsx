import React from 'react';

export const createRows = (data) =>
  data.map(({ name, status }) => ({
    cells: [name, status],
  }));
