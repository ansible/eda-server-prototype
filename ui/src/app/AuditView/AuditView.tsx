import React from 'react';
import { Level, LevelItem, Title } from '@patternfly/react-core';
import { Route, Switch, useLocation } from 'react-router-dom';
import { useIntl } from 'react-intl';
import AppTabs from '@app/shared/app-tabs';
import { TopToolbar } from '@app/shared/top-toolbar';
import { AnyObject, TabItemType } from '@app/shared/types/common-types';
import sharedMessages from '../messages/shared.messages';
import { AuditRules } from '@app/AuditView/audit-rules';
import { AuditHosts } from '@app/AuditView/audit-hosts';

export interface AuditRuleType {
  id: string;
  name: string;
  job: string;
  job_name: string;
  status: string;
  ruleset: string;
  ruleset_name: string;
  fired_at?: string;
}

export interface AuditHostType {
  rule: { id: string; name: string };
  job?: { id: string; name: string };
  ruleset?: { id: string; name: string };
  status: string;
  fired_date?: string;
}

const buildAuditTabs = (intl: AnyObject): TabItemType[] => [
  { eventKey: 0, title: intl.formatMessage(sharedMessages.audit_rules_title), name: `/audit/rules` },
  {
    eventKey: 1,
    title: intl.formatMessage(sharedMessages.audit_hosts_title),
    name: `/audit/hosts`,
  },
];

export const renderAuditTabs = (intl) => {
  const audit_tabs = buildAuditTabs(intl);
  return <AppTabs tabItems={audit_tabs} defaultActive={0} />;
};

const AuditView: React.FunctionComponent = () => {
  const intl = useIntl();

  return (
    <React.Fragment>
      <TopToolbar>
        <Level>
          <LevelItem>
            <Title headingLevel={'h2'}>{intl.formatMessage(sharedMessages.audit_view_title)}</Title>
          </LevelItem>
        </Level>
      </TopToolbar>
      <Switch>
        <Route exact path="/audit/hosts">
          <AuditHosts />
        </Route>
        <Route path="/audit">
          <AuditRules />
        </Route>
      </Switch>
    </React.Fragment>
  );
};

export { AuditView };
