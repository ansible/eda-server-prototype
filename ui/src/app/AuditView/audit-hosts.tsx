import { Button, PageSection } from '@patternfly/react-core';
import React, { useEffect, useReducer, useState } from 'react';
import { TableToolbarView } from '@app/shared/table-toolbar-view';
import sharedMessages from '../messages/shared.messages';
import { cellWidth } from '@patternfly/react-table';
import TableEmptyState from '@app/shared/table-empty-state';
import { useIntl } from 'react-intl';
import { defaultSettings } from '@app/shared/pagination';
import { createRows } from './audit-hosts-table-helpers';
import { CubesIcon } from '@patternfly/react-icons';
import { renderAuditTabs } from '@app/AuditView/AuditView';
import { AuditHostType } from '@app/AuditView/AuditView';
import { listAuditHosts } from '@app/API/Audit';

const columns = (intl) => [
  {
    title: intl.formatMessage(sharedMessages.name),
    transforms: [cellWidth(40)],
  },
  {
    title: intl.formatMessage(sharedMessages.rule),
  },
  {
    title: intl.formatMessage(sharedMessages.ruleset),
  },
  {
    title: intl.formatMessage(sharedMessages.lastFiredDate),
  },
];

const prepareChips = (filterValue, intl) =>
  filterValue
    ? [
        {
          category: intl.formatMessage(sharedMessages.name),
          key: 'name',
          chips: [{ name: filterValue, value: filterValue }],
        },
      ]
    : [];

const initialState = (filterValue = '') => ({
  filterValue,
  isFetching: false,
  isFiltering: false,
  rows: [],
});

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const auditHostsListState = (state, action) => {
  switch (action.type) {
    case 'setRows':
      return {
        ...state,
        rows: action.payload,
      };
    case 'setFetching':
      return {
        ...state,
        isFetching: action.payload,
      };
    case 'setFilterValue':
      return { ...state, filterValue: action.payload };
    case 'setFilteringFlag':
      return {
        ...state,
        isFiltering: action.payload,
      };
    case 'clearFilters':
      return { ...state, filterValue: '', isFetching: true };
    default:
      return state;
  }
};

const AuditHosts: React.FunctionComponent = () => {
  const [limit, setLimit] = useState(defaultSettings.limit);
  const [offset, setOffset] = useState(1);
  const [auditHosts, setAuditHosts] = useState<AuditHostType[]>([]);

  const meta = { count: auditHosts?.length || 0, limit, offset };

  const [{ filterValue, isFetching, isFiltering, rows }, stateDispatch] = useReducer(
    auditHostsListState,
    initialState()
  );

  const intl = useIntl();
  const updateAuditHosts = (pagination) => {
    stateDispatch({ type: 'setFetching', payload: true });
    return listAuditHosts(pagination)
      .then((data) => {
        setAuditHosts(data.data);
        return stateDispatch({ type: 'setFetching', payload: false });
      })
      .catch(() => stateDispatch({ type: 'setFetching', payload: false }));
  };

  useEffect(() => {
    stateDispatch({ type: 'setRows', payload: createRows(auditHosts) });
  }, [auditHosts]);

  useEffect(() => {
    updateAuditHosts(defaultSettings);
  }, []);

  const clearFilters = () => {
    stateDispatch({ type: 'clearFilters' });
    return updateAuditHosts(meta);
  };

  const handleFilterChange = (value) => {
    !value || value === '' ? clearFilters() : stateDispatch({ type: 'setFilterValue', payload: value });
  };

  return (
    <PageSection page-type={'activation-details'} id={'activation-details'}>
      {renderAuditTabs(intl)}
      <TableToolbarView
        ouiaId={'audit-hosts-table'}
        rows={rows}
        setLimit={setLimit}
        setOffset={setOffset}
        columns={columns(intl)}
        fetchData={updateAuditHosts}
        pagination={defaultSettings}
        plural={intl.formatMessage(sharedMessages.hosts)}
        singular={intl.formatMessage(sharedMessages.host)}
        isLoading={isFetching || isFiltering}
        onFilterChange={handleFilterChange}
        renderEmptyState={() => (
          <TableEmptyState
            title={intl.formatMessage(sharedMessages.nohostsrecentlychanged)}
            Icon={CubesIcon}
            PrimaryAction={() =>
              filterValue !== '' ? (
                <Button onClick={() => clearFilters()} variant="link">
                  {intl.formatMessage(sharedMessages.clearAllFilters)}
                </Button>
              ) : null
            }
            description={
              filterValue === ''
                ? intl.formatMessage(sharedMessages.nohostsrecentlychanged)
                : intl.formatMessage(sharedMessages.clearAllFiltersDescription)
            }
          />
        )}
      />
    </PageSection>
  );
};

export { AuditHosts };
