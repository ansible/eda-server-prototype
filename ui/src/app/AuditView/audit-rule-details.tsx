import {
  Card,
  CardBody,
  Flex,
  FlexItem,
  Grid,
  GridItem,
  PageSection,
  Stack,
  StackItem,
  Title,
  ToggleGroup,
  ToggleGroupItem,
} from '@patternfly/react-core';
import { Link } from 'react-router-dom';
import React, { useState, useEffect } from 'react';
import ReactJsonView from 'react-json-view';
import AceEditor from 'react-ace';
import 'ace-builds/src-noconflict/mode-yaml';
import 'ace-builds/src-noconflict/theme-xcode';
import styled from 'styled-components';
import { RuleType } from '@app/RuleSets/RuleSets';
import { ExtraVarType } from '@app/Vars/Vars';
import { useIntl } from 'react-intl';
import { fetchRuleVars } from '@app/API/Extravar';
import {renderAuditRuleTabs} from "@app/AuditView/AuditRule";

const FocusWrapper = styled.div`
  && + .keyboard-help-text {
    opacity: 0;
    transition: opacity 0.1s linear;
  }
  &:focus-within + .keyboard-help-text {
    opacity: 1;
  }
  & .ace_hidden-cursors .ace_cursor {
    opacity: 0;
  }
`;

const AuditRuleDetails: React.FunctionComponent<{ rule: RuleType }> = ({ rule }) => {
  const [ruleVars, setRuleVars] = useState<ExtraVarType | undefined>(undefined);
  const [varFormat, setVarFormat] = useState('yaml');
  const id = rule?.id;
  const intl = useIntl();

  useEffect(() => {
    rule?.extra_var_id
      ? fetchRuleVars(rule.extra_var_id).then((data) => setRuleVars(data?.data))
      : setRuleVars(undefined);
  }, [rule]);

  const handleToggleVarFormat = (_, event) => {
    const id = event.currentTarget.id;
    setVarFormat(id);
  };

  const renderFlexRuleDetails: React.FunctionComponent<RuleType> = (rule) => (
    <Stack hasGutter={true}>
      <StackItem>
        <Flex>
          <Flex direction={{ default: 'column' }} flex={{ default: 'flex_1' }}>
            <FlexItem>
              <Stack>
                <StackItem>
                  <Title headingLevel="h3">Name</Title>
                </StackItem>
                <StackItem>{rule?.name}</StackItem>
              </Stack>
            </FlexItem>
            <FlexItem>
              <Stack>
                <StackItem>
                  <Title headingLevel="h3">Action</Title>
                </StackItem>
                {rule && rule.action && (
                  <StackItem>
                    <Link to={'/jobs'}>{rule?.action ? Object.keys(rule?.action) : 'Link to action'}</Link>
                  </StackItem>
                )}
              </Stack>
            </FlexItem>
            <FlexItem>
              <Stack>
                <StackItem>
                  <Title headingLevel="h3">Last fired date</Title>
                </StackItem>
                <StackItem>{rule?.last_fired_date}</StackItem>
              </Stack>
            </FlexItem>
            <FlexItem>
              <Stack>
                <StackItem>
                  <Title headingLevel="h3">Last modified</Title>
                </StackItem>
                <StackItem>{rule?.modified_at}</StackItem>
              </Stack>
            </FlexItem>
          </Flex>
          <Flex direction={{ default: 'column' }} flex={{ default: 'flex_1' }}>
            <FlexItem>
              <Stack>
                <StackItem>
                  <Title headingLevel="h3">Description</Title>
                </StackItem>
                <StackItem>{rule?.description}</StackItem>
              </Stack>
            </FlexItem>
            <FlexItem>
              <Stack>
                <StackItem>
                  <Title headingLevel="h3">Project</Title>
                </StackItem>
                {rule && rule.project_id && (
                  <StackItem>
                    <Link to={'/project/' + rule?.project_id}>
                      {rule?.project_name || `Project ${rule?.project_id}`}
                    </Link>
                  </StackItem>
                )}
              </Stack>
            </FlexItem>
            <FlexItem>
              <Stack>
                <StackItem>
                  <Title headingLevel="h3">Fire count</Title>
                </StackItem>
                <StackItem>{rule?.fired_count || 0}</StackItem>
              </Stack>
            </FlexItem>
          </Flex>
          <Flex direction={{ default: 'column' }} flex={{ default: 'flex_1' }}>
            <FlexItem>
              <Stack>
                <StackItem>
                  <Title headingLevel="h3">Rule set</Title>
                </StackItem>
                {rule && rule.ruleset && (
                  <StackItem>
                    {<Link to={'/ruleset/' + rule.ruleset.id}>{rule.ruleset.name || rule.ruleset.name}</Link>}
                  </StackItem>
                )}
              </Stack>
            </FlexItem>
            <FlexItem>
              <Stack>
                <StackItem>
                  <Title headingLevel="h3">Rule type</Title>
                </StackItem>
                <StackItem>{rule?.type}</StackItem>
              </Stack>
            </FlexItem>
            <FlexItem>
              <Stack>
                <StackItem>
                  <Title headingLevel="h3">Created</Title>
                </StackItem>
                <StackItem>{rule?.created_at}</StackItem>
              </Stack>
            </FlexItem>
          </Flex>
        </Flex>
      </StackItem>
      <StackItem>
        <Stack hasGutter={true}>
          <StackItem>
            <Grid>
              <GridItem span={1}>
                <Title headingLevel="h3">Variables</Title>
              </GridItem>
              <GridItem span={2}>
                <ToggleGroup isCompact aria-label="JsonYaml">
                  <ToggleGroupItem
                    text="YAML"
                    buttonId="yaml"
                    isSelected={varFormat === 'yaml'}
                    onChange={handleToggleVarFormat}
                  />
                  <ToggleGroupItem
                    text="JSON"
                    buttonId="json"
                    isSelected={varFormat === 'json'}
                    onChange={handleToggleVarFormat}
                  />
                </ToggleGroup>
              </GridItem>
            </Grid>
          </StackItem>
          <StackItem>
            {ruleVars ? (
              varFormat === 'json' ? (
                <ReactJsonView displayObjectSize={false} displayDataTypes={false} quotesOnKeys={false} src={ruleVars} />
              ) : (
                <Card>
                  <FocusWrapper>
                    <AceEditor
                      mode="javascript"
                      theme="xcode"
                      name="rule_extravars"
                      fontSize={14}
                      value={ruleVars?.extra_var}
                      height={'100px'}
                      setOptions={{
                        enableBasicAutocompletion: false,
                        enableLiveAutocompletion: false,
                        enableSnippets: false,
                        showLineNumbers: true,
                        tabSize: 2,
                        readOnly: true,
                        focus: false,
                        highlightActiveLine: false,
                        cursorStart: 0,
                        cursorStyle: undefined,
                      }}
                    />
                  </FocusWrapper>
                </Card>
              )
            ) : null}
          </StackItem>
        </Stack>
      </StackItem>
    </Stack>
  );

  return (
    <PageSection page-type={'rule-details'} id={'rule-details'}>
      {renderAuditRuleTabs(id, intl)}
      <Stack>
        <StackItem>
          <Card>
            <CardBody>{renderFlexRuleDetails(rule)}</CardBody>
          </Card>
        </StackItem>
      </Stack>
    </PageSection>
  );
};

export { AuditRuleDetails };
