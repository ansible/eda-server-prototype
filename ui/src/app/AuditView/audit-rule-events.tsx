import { Button, PageSection } from '@patternfly/react-core';
import React, { useEffect, useReducer, useState } from 'react';
import { TableToolbarView } from '@app/shared/table-toolbar-view';
import sharedMessages from '../messages/shared.messages';
import TableEmptyState from '@app/shared/table-empty-state';
import { useIntl } from 'react-intl';
import { defaultSettings } from '@app/shared/pagination';
import { createRows } from './audit-rule-events-table-helpers';
import { CubesIcon } from '@patternfly/react-icons';
import {listAuditRuleEvents} from '@app/API/Audit';
import {RuleType} from "@app/Rules/Rules";
import {renderAuditRuleTabs} from "@app/AuditView/AuditRule";

export interface EventType {
  name: string;
  increment_counter?: string;
  source_type?: string;
  time_stamp?: string;
}

const columns = (intl) => [
  {
    title: intl.formatMessage(sharedMessages.name),
  },
  {
    title: intl.formatMessage(sharedMessages.increment_counter),
  },
  {
    title: intl.formatMessage(sharedMessages.source_type),
  },
  {
    title: intl.formatMessage(sharedMessages.timestamp),
  },
];

const prepareChips = (filterValue, intl) =>
  filterValue
    ? [
        {
          category: intl.formatMessage(sharedMessages.name),
          key: 'name',
          chips: [{ name: filterValue, value: filterValue }],
        },
      ]
    : [];

const initialState = (filterValue = '') => ({
  filterValue,
  isFetching: false,
  isFiltering: false,
  rows: [],
});

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const eventsListState = (state, action) => {
  switch (action.type) {
    case 'setRows':
      return {
        ...state,
        rows: action.payload,
      };
    case 'setFetching':
      return {
        ...state,
        isFetching: action.payload,
      };
    case 'setFilterValue':
      return { ...state, filterValue: action.payload };
    case 'setFilteringFlag':
      return {
        ...state,
        isFiltering: action.payload,
      };
    case 'clearFilters':
      return { ...state, filterValue: '', isFetching: true };
    default:
      return state;
  }
};

const AuditRuleEvents: React.FunctionComponent<{ rule: RuleType }> = ({ rule }) => {
  const [limit, setLimit] = useState(defaultSettings.limit);
  const [offset, setOffset] = useState(1);
  const [events, setEvents] = useState<EventType[]>([]);

  const meta = { count: events?.length || 0, limit, offset };
  const [{ filterValue, isFetching, isFiltering, rows }, stateDispatch] = useReducer(eventsListState, initialState());

  const intl = useIntl();
  const updateEvents = (pagination) => {
    stateDispatch({ type: 'setFetching', payload: true });
    return listAuditRuleEvents(rule?.id, pagination)
      .then(() => stateDispatch({ type: 'setFetching', payload: false }))
      .catch(() => stateDispatch({ type: 'setFetching', payload: false }));
  };

  useEffect(() => {
    updateEvents(defaultSettings);
  }, []);

  useEffect(() => {
    stateDispatch({ type: 'setRows', payload: createRows(events) });
  }, [events]);

  const clearFilters = () => {
    stateDispatch({ type: 'clearFilters' });
    return updateEvents(meta);
  };

  const handleFilterChange = (value) => {
    !value || value === '' ? clearFilters() : stateDispatch({ type: 'setFilterValue', payload: value });
  };

  return (
    <PageSection page-type={'audit-rule-events'} id={'audit-rule-events'}>
      {renderAuditRuleTabs(rule?.id, intl)}
      <TableToolbarView
        ouiaId={'audit-rule-events-table'}
        rows={rows}
        setLimit={setLimit}
        setOffset={setOffset}
        columns={columns(intl)}
        fetchData={updateEvents}
        pagination={defaultSettings}
        plural={intl.formatMessage(sharedMessages.events)}
        singular={intl.formatMessage(sharedMessages.event)}
        isLoading={isFetching || isFiltering}
        onFilterChange={handleFilterChange}
        renderEmptyState={() => (
          <TableEmptyState
            title={intl.formatMessage(sharedMessages.noauditruleevents)}
            Icon={CubesIcon}
            PrimaryAction={() =>
              filterValue !== '' ? (
                <Button onClick={() => clearFilters()} variant="link">
                  {intl.formatMessage(sharedMessages.clearAllFilters)}
                </Button>
              ) : null
            }
            description={
              filterValue === ''
                ? intl.formatMessage(sharedMessages.noauditruleevents)
                : intl.formatMessage(sharedMessages.clearAllFiltersDescription)
            }
          />
        )}
      />
    </PageSection>
  );
};

export { AuditRuleEvents };
