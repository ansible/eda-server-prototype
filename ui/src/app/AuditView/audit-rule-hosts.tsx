import { Button, PageSection } from '@patternfly/react-core';
import React, { useEffect, useReducer, useState } from 'react';
import { TableToolbarView } from '@app/shared/table-toolbar-view';
import sharedMessages from '../messages/shared.messages';
import { cellWidth } from '@patternfly/react-table';
import TableEmptyState from '@app/shared/table-empty-state';
import { useIntl } from 'react-intl';
import { defaultSettings } from '@app/shared/pagination';
import { createRows } from './audit-rule-hosts-table-helpers';
import { CubesIcon } from '@patternfly/react-icons';
import {listAuditRuleHosts} from '@app/API/Audit';
import {RuleType} from "@app/Rules/Rules";
import {renderAuditRuleTabs} from "@app/AuditView/AuditRule";

export interface HostType{
  name: string;
  status?: string;
}

const columns = (intl) => [
  {
    title: intl.formatMessage(sharedMessages.jobs),
    transforms: [cellWidth(40)],
  },
  {
    title: intl.formatMessage(sharedMessages.status),
  },
  {
    title: intl.formatMessage(sharedMessages.rule),
  },
  {
    title: intl.formatMessage(sharedMessages.lastFiredDate),
  },
];

const prepareChips = (filterValue, intl) =>
  filterValue
    ? [
        {
          category: intl.formatMessage(sharedMessages.name),
          key: 'name',
          chips: [{ name: filterValue, value: filterValue }],
        },
      ]
    : [];

const initialState = (filterValue = '') => ({
  filterValue,
  isFetching: false,
  isFiltering: false,
  rows: [],
});

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const hostsListState = (state, action) => {
  switch (action.type) {
    case 'setRows':
      return {
        ...state,
        rows: action.payload,
      };
    case 'setFetching':
      return {
        ...state,
        isFetching: action.payload,
      };
    case 'setFilterValue':
      return { ...state, filterValue: action.payload };
    case 'setFilteringFlag':
      return {
        ...state,
        isFiltering: action.payload,
      };
    case 'clearFilters':
      return { ...state, filterValue: '', isFetching: true };
    default:
      return state;
  }
};

const AuditRuleHosts: React.FunctionComponent<{ rule: RuleType }> = ({ rule }) => {
  const [limit, setLimit] = useState(defaultSettings.limit);
  const [offset, setOffset] = useState(1);
  const [hosts, setHosts] = useState<HostType[]>([]);

  const meta = { count: hosts?.length || 0, limit, offset };
  const [{ filterValue, isFetching, isFiltering, rows }, stateDispatch] = useReducer(hostsListState, initialState());

  const intl = useIntl();
  const updateHosts = (pagination) => {
    stateDispatch({ type: 'setFetching', payload: true });
    return listAuditRuleHosts(rule?.id, pagination)
      .then(() => stateDispatch({ type: 'setFetching', payload: false }))
      .catch(() => stateDispatch({ type: 'setFetching', payload: false }));
  };

  useEffect(() => {
    updateHosts(defaultSettings);
  }, []);

  useEffect(() => {
    stateDispatch({ type: 'setRows', payload: createRows(hosts) });
  }, [hosts]);

  const clearFilters = () => {
    stateDispatch({ type: 'clearFilters' });
    return updateHosts(meta);
  };

  const handleFilterChange = (value) => {
    !value || value === '' ? clearFilters() : stateDispatch({ type: 'setFilterValue', payload: value });
  };

  return (
    <PageSection page-type={'audit-rule-hosts'} id={'audit-rule-hosts'}>
      {renderAuditRuleTabs(rule?.id, intl)}
      <TableToolbarView
        ouiaId={'audit-rule-hosts-table'}
        rows={rows}
        setLimit={setLimit}
        setOffset={setOffset}
        columns={columns(intl)}
        fetchData={updateHosts}
        pagination={defaultSettings}
        plural={intl.formatMessage(sharedMessages.hosts)}
        singular={intl.formatMessage(sharedMessages.host)}
        isLoading={isFetching || isFiltering}
        onFilterChange={handleFilterChange}
        renderEmptyState={() => (
          <TableEmptyState
            title={intl.formatMessage(sharedMessages.noauditrulehosts)}
            Icon={CubesIcon}
            PrimaryAction={() =>
              filterValue !== '' ? (
                <Button onClick={() => clearFilters()} variant="link">
                  {intl.formatMessage(sharedMessages.clearAllFilters)}
                </Button>
              ) : null
            }
            description={
              filterValue === ''
                ? intl.formatMessage(sharedMessages.noauditrulehosts)
                : intl.formatMessage(sharedMessages.clearAllFiltersDescription)
            }
          />
        )}
      />
    </PageSection>
  );
};

export { AuditRuleHosts };
