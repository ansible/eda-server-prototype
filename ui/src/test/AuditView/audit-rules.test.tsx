import * as React from 'react';
import { mount } from 'enzyme';
import { AuditHosts } from '@app/AuditView/audit-hosts';
import { MemoryRouter } from 'react-router';
import { act } from 'react-dom/test-utils';
import { IntlProvider } from 'react-intl';
import { Route } from 'react-router-dom';
import { mockApi } from '../__mocks__/baseApi';

const ComponentWrapper = ({ children, initialEntries = ['/dashboard'] }) => (
  <IntlProvider locale="en">
    <MemoryRouter initialEntries={initialEntries} keyLength={0} key={'test'}>
      {children}
    </MemoryRouter>
  </IntlProvider>
);

describe('AuditHosts', () => {
  beforeAll(() => {
    mockApi.reset();
  });

  it('should render the AuditView component', async () => {
    mockApi.onGet(`/api/audit/hosts_changed`).replyOnce(200, [
      { id: '1', name: 'Host 1' },
      { id: '2', name: 'Host 2' },
      { id: '3', name: 'Host 3' },
    ]);

    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper initialEntries={['/audit']}>
          <Route path="/audit/rules">
            <AuditHosts />
          </Route>
        </ComponentWrapper>
      );
    });
    await act(async () => {
      wrapper.update();
    });
    expect(wrapper).toMatchSnapshot();
  });
});
