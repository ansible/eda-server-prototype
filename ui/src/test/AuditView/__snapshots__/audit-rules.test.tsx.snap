// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuditHosts should render the AuditView component 1`] = `
<ComponentWrapper
  initialEntries={
    [
      "/audit",
    ]
  }
>
  <IntlProvider
    defaultFormats={{}}
    defaultLocale="en"
    fallbackOnEmptyString={true}
    formats={{}}
    locale="en"
    messages={{}}
    onError={[Function]}
    onWarn={[Function]}
    textComponent={Symbol(react.fragment)}
  >
    <MemoryRouter
      initialEntries={
        [
          "/audit",
        ]
      }
      key="test"
      keyLength={0}
    >
      <Router
        history={
          {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": [
              {
                "hash": "",
                "pathname": "/audit",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": {
              "hash": "",
              "pathname": "/audit",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <Route
          path="/audit/rules"
        />
      </Router>
    </MemoryRouter>
  </IntlProvider>
</ComponentWrapper>
`;
